File: src/Ast.py
from __future__ import annotations
from typing import List
from enum import Enum, auto
from utils import *
from Token import TokenKind, Span


def get_ty(name) -> PrimTyKind | None:
    match name:
        case "u8": return PrimTyKind.U8
        case "u16": return PrimTyKind.U16
        case "u32": return PrimTyKind.U32
        case "u64": return PrimTyKind.U64
        case "usize": return PrimTyKind.USize
        case "i8": return PrimTyKind.I8
        case "i16": return PrimTyKind.I16
        case "i32": return PrimTyKind.I32
        case "i64": return PrimTyKind.I64
        case "isize": return PrimTyKind.ISize
        case "f32": return PrimTyKind.F32
        case "f64": return PrimTyKind.F64
        case "bool": return PrimTyKind.Bool
        case "char": return PrimTyKind.Char
        case "str": return PrimTyKind.Str
        case "raw": return PrimTyKind.Raw
        case "()": return PrimTyKind.Unit


class Ty:
    pass


class PrimTyKind(Enum):
    U8 = auto()
    U16 = auto()
    U32 = auto()
    U64 = auto()
    USize = auto()
    I8 = auto()
    I16 = auto()
    I32 = auto()
    I64 = auto()
    ISize = auto()
    F32 = auto()
    F64 = auto()
    Bool = auto()
    Char = auto()
    Str = auto()
    Raw = auto()
    Unit = auto()

    def get_size(self) -> int:
        match self:
            case PrimTyKind.U8: return 1
            case PrimTyKind.U16: return 2
            case PrimTyKind.U32: return 4
            case PrimTyKind.U64: return 8
            case PrimTyKind.USize: return 8
            case PrimTyKind.I8: return 1
            case PrimTyKind.I16: return 2
            case PrimTyKind.I32: return 4
            case PrimTyKind.I64: return 8
            case PrimTyKind.ISize: return 8
            case PrimTyKind.F32: return 4
            case PrimTyKind.F64: return 8
            case PrimTyKind.Bool: return 1
            case PrimTyKind.Char: return 1
            case PrimTyKind.Str: return 8
            case PrimTyKind.Raw: return 8
            case PrimTyKind.Unit: return 0

    def is_number(self) -> bool:
        match self:
            case PrimTyKind.Bool \
                | PrimTyKind.Char \
                | PrimTyKind.Str \
                | PrimTyKind.Unit \
                | PrimTyKind.Raw: return False
            case _: return True

    def is_float(self) -> bool:
        match self:
            case PrimTyKind.F32 | PrimTyKind.F64: return True
        return False

    def is_int(self) -> bool:
        match self:
            case PrimTyKind.Bool \
                | PrimTyKind.Char \
                | PrimTyKind.Str \
                | PrimTyKind.Unit \
                | PrimTyKind.F32 \
                | PrimTyKind.F64 \
                | PrimTyKind.Raw: return False
        return True

    def __repr__(self) -> str:
        match self:
            case PrimTyKind.U8: return "u8"
            case PrimTyKind.U16: return "u16"
            case PrimTyKind.U32: return "u32"
            case PrimTyKind.U64: return "u64"
            case PrimTyKind.USize: return "usize"
            case PrimTyKind.I8: return "i8"
            case PrimTyKind.I16: return "i16"
            case PrimTyKind.I32: return "i32"
            case PrimTyKind.I64: return "i64"
            case PrimTyKind.ISize: return "isize"
            case PrimTyKind.F32: return "f32"
            case PrimTyKind.F64: return "f64"
            case PrimTyKind.Bool: return "bool"
            case PrimTyKind.Char: return "char"
            case PrimTyKind.Str: return "str"
            case PrimTyKind.Raw: return "raw"
            case PrimTyKind.Unit: return "()"


class PrimTy(Ty):
    __match_args__ = ("kind", )

    def __init__(self, kind: PrimTyKind):
        self.kind = kind
        self.span: Span | None = None

    def is_int(self) -> bool:
        return self.kind.is_int()

    def is_float(self) -> bool:
        return self.kind.is_float()

    def get_size(self) -> int:
        return self.kind.get_size()

    def __eq__(self, __o: Ty) -> bool:
        match __o:
            case PrimTy(kind):
                return kind == self.kind
            case _:
                return False

    def __ne__(self, __o: Ty) -> bool:
        return not self.__eq__(__o)

    def __repr__(self) -> str:
        return repr(self.kind)


class RefTy(Ty):
    __match_args__ = ("ref", )

    def __init__(self, ref: Ty):
        self.ref = ref
        self.span = None

    def get_size(self) -> int:
        return 8

    def __eq__(self, __o: Ty) -> bool:
        match __o:
            case RefTy(ref):
                return ref == self.ref
            case _:
                return False

    def __ne__(self, __o: Ty) -> bool:
        return not self.__eq__(__o)

    def __repr__(self) -> str:
        return f"&{repr(self.ref)}"


class FnArg:
    pass


class Arg(FnArg):
    __match_args__ = ("name", "ty", )

    def __init__(self, name, ty):
        self.name = name
        self.ty = ty
        self.span: Span | None = None


class Variadic(FnArg):
    pass


class Fn:
    __match_args__ = ("name", "args", "ret_ty", "body", "is_extern", "abi", )

    def __init__(
        self,
        name: str,
        args: List[FnArg],
        ret_ty: Ty | None,
        body: Block | None,
        is_extern: bool,
        abi: str | None
    ):
        self.name = name
        self.args = args
        self.ret_ty = ret_ty
        self.body = body
        self.is_extern = is_extern
        self.abi = abi
        self.alignment = ((self.local_count + 1) // 2) * 16
        self.span: Span | None = None

    def aligned_add(self, off, size):
        off += size
        return off

    @property
    def stack_offset(self):
        stack_off = 0
        for arg in self.args:
            match arg:
                case Arg(_, ty):
                    size = ty.get_size()
                    stack_off = self.aligned_add(stack_off, size)
        if not self.body:
            return stack_off

        stack_off += self.body.calc_stack()
        return stack_off

    @property
    def stack_alignment(self):
        return (self.stack_offset + 15) & ~15

    @property
    def local_count(self):
        local_count_ = len(self.args)
        if not self.body:
            return local_count_
        for node in self.body.stmts:
            if type(node) == Let:
                local_count_ += 1
        return local_count_


class Block:
    __match_args__ = ("stmts", )

    def __init__(self, stmts: List[Stmt]):
        self.stmts = stmts
        self.span: Span | None = None

    def calc_stack(self) -> int:
        off = 0
        for stmt in self.stmts:
            match stmt.kind:
                case Let(_, ty, _):
                    size = ty.get_size()
                    off += size
                case Expr(kind):
                    match kind:
                        case Loop(block):
                            off += block.calc_stack()
                        case If(_, if_block, else_block):
                            off += if_block.calc_stack()
                            if else_block:
                                off += else_block.calc_stack()
        return off


class Expr:
    __match_args__ = ("kind", )

    def __init__(self, kind) -> None:
        self.kind = kind
        self.ty: Ty | None = None
        self.span: Span | None = None


class BinaryKind(Enum):
    Add = auto()
    Sub = auto()
    Mul = auto()
    Div = auto()
    Lt = auto()
    Gt = auto()
    Mod = auto()


def binary_kind_from_token(kind: TokenKind) -> BinaryKind:
    match kind:
        case TokenKind.PLUS: return BinaryKind.Add
        case TokenKind.MINUS: return BinaryKind.Sub
        case TokenKind.STAR: return BinaryKind.Mul
        case TokenKind.SLASH: return BinaryKind.Div
        case TokenKind.GT: return BinaryKind.Gt
        case TokenKind.LT: return BinaryKind.Lt
        case TokenKind.PERCENT: return BinaryKind.Mod
        case _: assert False


class Binary:
    __match_args__ = ("kind", "left", "right")

    def __init__(self, kind: BinaryKind, left: Expr, right: Expr):
        self.kind = kind
        self.left = left
        self.right = right
        self.ty = None
        self.span: Span | None = None


class UnaryKind(Enum):
    Neg = auto()
    Not = auto()
    AddrOf = auto()


def unary_kind_from_token(kind: TokenKind) -> UnaryKind:
    match kind:
        case TokenKind.MINUS: return UnaryKind.Neg
        case TokenKind.BANG: return UnaryKind.Not
        case TokenKind.AMPERSAND: return UnaryKind.AddrOf
        case _: assert False


class Unary:
    __match_args__ = ("kind", "expr", )

    def __init__(self, kind: UnaryKind, expr: Expr):
        self.kind = kind
        self.expr = expr
        self.ty = None
        self.span: Span | None = None


class Call:
    __match_args__ = ("name", "args", )

    def __init__(self, name, args: List[Expr]):
        self.name = name
        self.args = args
        self.ty = None
        self.span: Span | None = None


class If:
    __match_args__ = ("cond", "body", "elze", )

    def __init__(self, cond, body, elze=None):
        self.cond = cond
        self.body = body
        self.elze = elze
        self.ty = None
        self.span: Span | None = None


class Else:
    __match_args__ = ("body", )

    def __init__(self, body):
        self.body = body
        self.span: Span | None = None


class Loop:
    __match_args__ = ("body", )

    def __init__(self, body: Block) -> None:
        self.body = body
        self.span: Span | None = None


class Break:
    def __init__(self) -> None:
        self.label = 0


class Stmt:
    def __init__(self, kind) -> None:
        self.kind = kind
        self.span: Span | None = None


class Let:
    __match_args__ = ("name", "ty", "init")

    def __init__(self, name, ty, init):
        self.name = name
        self.ty = ty
        self.init = init
        self.span: Span | None = None


class Assign:
    __match_args__ = ("name", "init", )

    def __init__(self, name, init):
        self.name = name
        self.init = init
        self.span: Span | None = None


class Ident:
    __match_args__ = ("name", )

    def __init__(self, name):
        self.name = name
        self.ty = None
        self.span: Span | None = None


class Cast:
    __match_args__ = ("expr", "ty", )

    def __init__(self, expr, ty):
        self.expr = expr
        self.ty = ty
        self.span: Span | None = None


class Lit(Enum):
    Int = auto()
    Float = auto()
    Bool = auto()
    Str = auto()
    Char = auto()


class Literal:
    __match_args__ = ("kind", "value", )

    def __init__(self, kind, value):
        self.kind = kind
        self.value = value
        self.ty = None
        self.span: Span | None = None

File: ina/ast.ml
open Token
open Source
open Structures.Vec

type node_id = int
type ident = string

type int_ty =
  | I8
  | I16
  | I32
  | I64
  | Isize
  | U8
  | U16
  | U32
  | U64
  | Usize

type float_ty =
  | F64
  | F32

type ty_kind =
  | Int of int_ty
  | Float of float_ty
  | Bool
  | Str
  | Ptr of (mutability * ty)
  | Ref of (mutability * ty)
  | Slice of ty
  | FnPtr of ty vec * ty * bool
  | Path of path
  | ImplicitSelf
  | Unit
  | CVarArgs
  | Err

and mutability =
  | Mut
  | Imm

and path_segment = {
    ident: ident
  ; args: ty vec option
  ; span: Span.t
  ; id: node_id
}

and path = {
    segments: path_segment vec
  ; span: Span.t
  ; path_id: node_id
}

and ty = {
    kind: ty_kind
  ; span: Span.t
  ; ty_id: node_id
}

let mk_ty kind span ty_id = { kind; span; ty_id }
let is_self ty = match ty.kind with ImplicitSelf -> true | _ -> false

type attr = {
    kind: attr_kind
  ; style: attr_style
  ; attr_span: Span.t
  ; attr_id: node_id
}

and attr_vec = attr vec

and attr_kind =
  | Doc of string
  | NormalAttr of normal_attr

and normal_attr = { name: ident }

and attr_style =
  | Outer
  | Inner

let is_test attrs =
  find
    (function
     | { kind = NormalAttr { name = "test" }; _ } as attr -> Some attr
     | _ -> None)
    attrs
  |> Option.is_some
;;

type modd = {
    mutable items: item vec
  ; mutable attrs: attr vec
  ; mutable mod_name: string
  ; mod_path: string
  ; mod_span: Span.t
  ; mod_id: node_id
}

and generic_param_kind = Ident of ident

and generic_param = {
    kind: generic_param_kind
  ; span: Span.t
  ; id: node_id
}

and generics = {
    params: generic_param vec
  ; span: Span.t
  ; id: node_id
}

and strukt = {
    name: string
  ; mutable fields: (ty * string) vec
  ; generics: generics
  ; span: Span.t
  ; id: node_id
}

and variant = {
    name: string
  ; fields: ty vec
  ; span: Span.t
  ; id: node_id
}

and adt = {
    name: string
  ; variants: variant vec
  ; generics: generics
  ; span: Span.t
  ; id: node_id
}

and typ =
  | Struct of strukt
  | Adt of adt

and assoc_item = AssocFn of func

and impl = {
    ty: ty
  ; generics: generics
  ; items: assoc_item vec
  ; span: Span.t
  ; id: node_id
}

and using_kind =
  | Simple of ident option
  | Val of ident * Span.t
  | Nested of using vec
  | Glob

and using = {
    prefix: path
  ; kind: using_kind
  ; span: Span.t
  ; id: node_id
}

and item =
  | Fn of func * attr_vec
  | Type of typ
  | Foreign of func vec * node_id
  | Impl of impl
  | Mod of {
        name: string
      ; mutable resolved_mod: modd option
      ; inline: bool
      ; span: Span.t
    }
  | ExternMod of string
  | Using of using

and arg = {
    ty: ty
  ; arg: ident
  ; arg_id: node_id
}

and fn_sig = {
    mutable args: arg vec
  ; ret_ty: ty option
  ; fn_span: Span.t
  ; is_variadic: bool
  ; fn_sig_id: node_id
}

and func = {
    is_extern: bool
  ; name: ident
  ; abi: string
  ; fn_sig: fn_sig
  ; fn_generics: generics
  ; body: block option
  ; func_span: Span.t
  ; func_id: node_id
}

and block = {
    block_stmts: stmt vec
  ; last_expr: expr option
  ; block_span: Span.t
  ; block_id: node_id
}

and stmt =
  | Stmt of expr
  | Expr of expr
  | Assign of expr * expr
  | Binding of binding
  | Assert of expr * expr option

and binding = {
    binding_pat: pat
  ; mutable binding_ty: ty option
  ; binding_expr: expr
  ; binding_id: node_id
  ; binding_span: Span.t
}

and pat =
  | PIdent of (mutability * ident * node_id ref)
  | PPath of path
  | PCons of (path * pat vec)
  | POr of pat vec
  | PRange of (lit * lit)
  | PLit of lit
  | PWild

and expr = {
    mutable expr_kind: expr_kind
  ; expr_span: Span.t
  ; expr_id: node_id
}

and binary_kind =
  | Add
  | Sub
  | Mul
  | Div
  | Gt
  | Lt
  | Eq
  | NotEq
  | GtEq
  | LtEq
  | BitAnd
  | BitOr
  | And
  | Or

and expr_kind =
  | Lit of lit
  | Path of path
  | Call of expr * expr vec
  | Binary of binary_kind * expr * expr
  | If of iff
  | Block of block
  | Deref of expr
  | Ref of (mutability * expr)
  | StructExpr of struct_expr
  | Field of expr * ident
  | Cast of expr * ty
  | MethodCall of expr * path_segment * expr vec
  | Match of expr * arm vec
  | Slice of expr vec
  | Index of (expr * expr)
  | Hole

and arm = {
    pat: pat
  ; patspan: Span.t
  ; expr: expr
  ; span: Span.t
}

and struct_expr = {
    struct_name: path
  ; fields: (string * expr) vec
  ; struct_expr_span: Span.t
  ; struct_expr_id: node_id
}

and iff = {
    cond: expr
  ; then_block: block
  ; else_block: expr option
  ; if_span: Span.t
  ; if_id: node_id
}

and lit =
  | LitInt of int
  | LitFloat of float
  | LitBool of bool
  | LitStr of string

let binary_kind_from_token = function
  | Plus -> Add
  | Minus -> Sub
  | Star -> Mul
  | Slash -> Div
  | EqEq -> Eq
  | BangEq -> NotEq
  | Ampersand -> BitAnd
  | Pipe -> BitOr
  | Ampersand2 -> And
  | Pipe2 -> Or
  | LAngle -> Lt
  | RAngle -> Gt
  | _ -> assert false
;;
